src/TriggerBinding/SqlTableChangeMonitor.cs

```diff
@@ -842,44 +860,53 @@ private SqlCommand BuildGetChangesCommand(SqlConnection connection, SqlTransacti
         }
 
         /// <summary>
-        /// Returns the number of changes(rows) on the user's table that are actively locked by other leases OR returns -1 on exception.
+        /// Returns a message indicating the number of changes(rows) on the user's table that are actively locked by other leases or have
+        /// reached the max attempts allowed for the row.
         /// </summary>
         /// <param name="connection">The connection to add to the SqlCommand</param>
         /// <param name="transaction">The transaction to add to the SqlCommand</param>
-        /// <returns>The number of rows locked by leases or -1 on exception</returns>
-        private async Task<int> GetLeaseLockedRowCount(SqlConnection connection, SqlTransaction transaction)
+        /// <param name="token">Cancellation token</param>
+        /// <returns>The message with the number of rows that were lease locked or were at the max attempt limit. Null if no such rows exist</returns>
+        private async Task<string> GetLeaseLockedOrMaxAttemptRowCountMessage(SqlConnection connection, SqlTransaction transaction, CancellationToken token)
         {
             string leasesTableJoinCondition = string.Join(" AND ", this._primaryKeyColumns.Select(col => $"c.{col.name.AsBracketQuotedString()} = l.{col.name.AsBracketQuotedString()}"));
-            int leaseLockedRowsCount = 0;
-            long getLockedRowCountDurationMs = 0L;
             // Get the count of changes from CHANGETABLE that meet the following criteria:
-            // * Not Null LeaseExpirationTime AND
-            // * LeaseExpirationTime > Current Time
-            string getLeaseLockedrowCountQuery = $@"
+            // Lease locked:
+            //  * Have attempts remaining (Attempt count < Max attempts)
+            //  * NOT NULL LeaseExpirationTime
+            //  * LeaseExpirationTime > Current Time
+            // Max Attempts reached:
+            //  * NULL LeaseExpirationTime OR LeaseExpirationTime <= Current Time
+            //  * No attempts remaining (Attempt count = Max attempts)
+            string getLeaseLockedOrMaxAttemptRowCountQuery = $@"
                 {AppLockStatements}
 
                 DECLARE @last_sync_version bigint;
                 SELECT @last_sync_version = LastSyncVersion
                 FROM {GlobalStateTableName}
                 WHERE UserFunctionID = '{this._userFunctionId}' AND UserTableID = {this._userTableId};
-
-                SELECT COUNT(*)
+                DECLARE @lease_locked_count int;
```

Comment:
Refactored this to just create the message in the query itself. I could have returned a table with the numeric values, but having to read the data table and do all that extra stuff didn't seem worth it when we can just do it directly in the T-SQL.